coax/td_learning/_clippeddoubleqlearning.py

```diff
@@ -141,12 +141,20 @@ def __init__(
         self._super_grads_and_metrics_func = self._grads_and_metrics_func
         self._super_td_error_func = self._td_error_func
 
+        def make_batch(examples):
+            return jax.tree_map(lambda *a: jnp.stack(a), *examples)
+
         def q_targ_pi_argmin(target_params, target_state, rng, transition_batch):
             rngs = hk.PRNGSequence(rng)
 
-            def q_targ_target_func(p, s):
-                return self.q_targ_func(p, s, rng=next(rngs), transition_batch=transition_batch)
-            targets = jax.vmap(q_targ_target_func)(*stack_trees(target_params, target_state))
+            vq_targ_func = jax.vmap(lambda p, s: self.q_targ_func(
+                p, s, rng=next(rngs), transition_batch=transition_batch),
+                in_axes=({'q_targ': 0, 'pi_targ': None}, {'q_targ': 0, 'pi_targ': None}))
+            vvq_targ_func = jax.vmap(vq_targ_func, in_axes=(
+                {'q_targ': None, 'pi_targ': 0}, {'q_targ': None, 'pi_targ': 0}))
+            targets = vvq_targ_func(make_batch(target_params),
+                                    make_batch(target_state))
+            targets = targets.reshape(len(self.q_targ_list) * len(self.pi_targ_list), -1)
```

Comment:
This is how I imagine it should work, but it throws an error that the value does not have the prefix that I provided in `in_axes`. The simple cases all work but as soon as the parameters and state get involved it fails. I will have to investigate this further.

Reply:
I think the tree structure of in/out_axes must match those of it's input (up to the level you provide). So you'll probably have to transform them using `hk.data_structures.to_haiku_dict()` and include all keys, i.e. also 'q'.

Reply:
So something like

```python
in_axes = (
  hk.data_structures.to_haiku_dict({
    'q': None, 'q_targ': 0, 'pi_targ': None}),  # params
  hk.data_structures.to_haiku_dict({
    'q': None, 'q_targ': 0, 'pi_targ': None}),  # function_state
  None,  # rng
  ...)  # etc.
```